yt_dlp/utils.py

```diff
@@ -5284,107 +5285,174 @@ def load_plugins(name, suffix, namespace):
                 key = ...
 
-            if key is ...:
-                obj = (obj.values() if isinstance(obj, dict)
-                       else obj if isinstance(obj, (list, tuple, LazyList))
-                       else str(obj) if traverse_string else [])
-                _current_depth += 1
-                depth = max(depth, _current_depth)
-                return [_traverse_obj(inner_obj, path[i + 1:], _current_depth) for inner_obj in obj]
-            elif isinstance(key, dict):
-                obj = filter_dict({k: _traverse_obj(obj, v, _current_depth) for k, v in key.items()})
-            elif callable(key):
-                if isinstance(obj, (list, tuple, LazyList)):
-                    obj = enumerate(obj)
+            if not casesense and isinstance(key, str):
+                key = key.casefold()
+
+            result_objs.clear()
+            for obj in objs:
+                if key is None:
+                    result_objs.append(obj)
+                    continue
+                elif obj is None:
+                    continue
+
+                if isinstance(key, (list, tuple)):
+                    has_branched = True
+                    for branch in key:
+                        _, result = _traverse_path(obj, branch)
+                        result_objs.extend(result)
+
+                elif key is ...:
+                    has_branched = True
+                    if isinstance(obj, collections.abc.Mapping):
+                        result_objs.extend(obj.values())
+                    elif is_sequence(obj):
+                        result_objs.extend(obj)
+                    elif traverse_string:
+                        result_objs.extend(str(obj))
+
+                elif callable(key):
+                    has_branched = True
+                    if is_sequence(obj):
+                        iter_obj = enumerate(obj)
+                    elif isinstance(obj, collections.abc.Mapping):
+                        iter_obj = obj.items()
+                    elif traverse_string:
+                        iter_obj = enumerate(str(obj))
+                    else:
+                        iter_obj = ()
+                    result_objs.extend(v for k, v in iter_obj if try_call(key, args=(k, v)))
+
+                elif isinstance(key, dict):
+                    obj = [(k, _traverse_obj(obj, v)) for k, v in key.items()]
+                    result_objs.append({k: v if v is not None else default for k, v in obj
+                                        if v is not None or (v is None and default is not None)})
+
                 elif isinstance(obj, dict):
-                    obj = obj.items()
+                    if casesense:
+                        result = obj.get(key)
+                    else:
+                        result = next((v for k, v in obj.items() if _casefold(k) == key), None)
+                    if result is not None:
+                        result_objs.append(result)
```

Comment:
since we filter out `None` below, do we need to check it here? If not, cant we just directly do
```py
                    result_objs.append(next(...))
```

Reply:
I tried to avoid the unneccesary function call, adding the item just to filter it out later doesnt make much sense. I don't think it hinders readability, although I am also not opposed to removing the `if` if required.

On the contrary, if removing the check, it might also make sense to get rid of the filtering completely:
```py
for obj in objs:
    if obj is None:
        continue
    ...
    objs = result_objs

return has_branched, [obj for obj in objs if obj is not None]
```